File: modulo03-mcp-na-pratica/mcp-customers-server-java/src/main/java/com/iadeva/mcp/client/CustomerHttpClient.java
package com.iadeva.mcp.client;

// Cliente HTTP para a legacy customer API — abstrai todas as chamadas REST para a API legada
// Equivalente ao httpClient.ts / fetch calls nos projetos TypeScript do curso

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.iadeva.mcp.model.Customer;
import com.iadeva.mcp.model.MutationResult;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Componente responsável por toda comunicação HTTP com a legacy customer API.
 * Autentica via Bearer token lido da variável de ambiente API_SERVICE_TOKEN.
 */
@Component
public class CustomerHttpClient {

    private static final Logger logger = LoggerFactory.getLogger(CustomerHttpClient.class);

    private final RestClient restClient;
    private final ObjectMapper objectMapper;

    public CustomerHttpClient(
            @Value("${app.legacy-api.base-url:http://localhost:3000}") String baseUrl,
            @Value("${app.legacy-api.service-token:}") String serviceToken,
            ObjectMapper objectMapper
    ) {
        this.objectMapper = objectMapper;
        // Configura RestClient com base URL e header de autenticação padrão
        this.restClient = RestClient.builder()
                .baseUrl(baseUrl)
                .defaultHeader("Authorization", "Bearer " + serviceToken)
                .defaultHeader("Content-Type", "application/json")
                .build();

        logger.info("CustomerHttpClient inicializado com base URL: {}", baseUrl);
    }

    /**
     * Lista todos os clientes cadastrados na API legada.
     *
     * @return lista de clientes
     */
    public List<Customer> listAll() {
        try {
            String responseBody = restClient.get()
                    .uri("/customers")
                    .retrieve()
                    .body(String.class);

            return parseCustomerList(responseBody);
        } catch (Exception e) {
            logger.error("Erro ao listar clientes: {}", e.getMessage());
            throw new RuntimeException("Falha ao listar clientes da API legada", e);
        }
    }

    /**
     * Busca um cliente específico pelo ID.
     *
     * @param id identificador do cliente
     * @return cliente encontrado ou null se não existir
     */
    public Customer findById(String id) {
        try {
            String responseBody = restClient.get()
                    .uri("/customers/{id}", id)
                    .retrieve()
                    .body(String.class);

            return parseCustomer(responseBody);
        } catch (Exception e) {
            logger.error("Erro ao buscar cliente por ID {}: {}", id, e.getMessage());
            return null;
        }
    }

    /**
     * Cria um novo cliente na API legada.
     *
     * @param name  nome do cliente
     * @param phone telefone do cliente
     * @return resultado da operação de criação
     */
    public MutationResult create(String name, String phone) {
        try {
            Map<String, String> payload = Map.of("name", name, "phone", phone);
            String body = objectMapper.writeValueAsString(payload);

            String responseBody = restClient.post()
                    .uri("/customers")
                    .body(body)
                    .retrieve()
                    .body(String.class);

            Customer created = parseCustomer(responseBody);
            return new MutationResult(true, "Cliente criado com sucesso", created);
        } catch (Exception e) {
            logger.error("Erro ao criar cliente: {}", e.getMessage());
            return new MutationResult(false, "Falha ao criar cliente: " + e.getMessage(), null);
        }
    }

    /**
     * Atualiza os dados de um cliente existente.
     *
     * @param id    identificador do cliente
     * @param name  novo nome (pode ser null para não alterar)
     * @param phone novo telefone (pode ser null para não alterar)
     * @return resultado da operação de atualização
     */
    public MutationResult update(String id, String name, String phone) {
        try {
            // Monta payload apenas com campos fornecidos
            Map<String, String> payload;
            if (name != null && phone != null) {
                payload = Map.of("name", name, "phone", phone);
            } else if (name != null) {
                payload = Map.of("name", name);
            } else if (phone != null) {
                payload = Map.of("phone", phone);
            } else {
                return new MutationResult(false, "Nenhum campo fornecido para atualização", null);
            }

            String body = objectMapper.writeValueAsString(payload);

            String responseBody = restClient.put()
                    .uri("/customers/{id}", id)
                    .body(body)
                    .retrieve()
                    .body(String.class);

            Customer updated = parseCustomer(responseBody);
            return new MutationResult(true, "Cliente atualizado com sucesso", updated);
        } catch (Exception e) {
            logger.error("Erro ao atualizar cliente {}: {}", id, e.getMessage());
            return new MutationResult(false, "Falha ao atualizar cliente: " + e.getMessage(), null);
        }
    }

    /**
     * Remove um cliente da API legada pelo ID.
     *
     * @param id identificador do cliente a ser removido
     * @return resultado da operação de exclusão
     */
    public MutationResult delete(String id) {
        try {
            restClient.delete()
                    .uri("/customers/{id}", id)
                    .retrieve()
                    .toBodilessEntity();

            return new MutationResult(true, "Cliente removido com sucesso", null);
        } catch (Exception e) {
            logger.error("Erro ao remover cliente {}: {}", id, e.getMessage());
            return new MutationResult(false, "Falha ao remover cliente: " + e.getMessage(), null);
        }
    }

    // --- Métodos auxiliares de parsing ---

    private Customer parseCustomer(String json) {
        try {
            JsonNode node = objectMapper.readTree(json);
            return new Customer(
                    node.path("id").asText(null),
                    node.path("name").asText(null),
                    node.path("phone").asText(null)
            );
        } catch (Exception e) {
            throw new RuntimeException("Erro ao parsear resposta de cliente: " + json, e);
        }
    }

    private List<Customer> parseCustomerList(String json) {
        try {
            JsonNode array = objectMapper.readTree(json);
            List<Customer> customers = new ArrayList<>();
            if (array.isArray()) {
                for (JsonNode node : array) {
                    customers.add(new Customer(
                            node.path("id").asText(null),
                            node.path("name").asText(null),
                            node.path("phone").asText(null)
                    ));
                }
            }
            return customers;
        } catch (Exception e) {
            throw new RuntimeException("Erro ao parsear lista de clientes: " + json, e);
        }
    }
}

File: modulo03-mcp-na-pratica/mcp-customers-server-go/client/customer_http_client.go
// Package client fala HTTP com a legacy customer API — equivalente a
// CustomerHttpClient.java.
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"mcp-customers-server/model"
)

// CustomerHttpClient é responsável por toda comunicação HTTP com a legacy
// customer API. Autentica via Bearer token.
type CustomerHttpClient struct {
	BaseURL      string
	ServiceToken string
	HTTPClient   *http.Client
}

// NewCustomerHttpClient cria um CustomerHttpClient com timeout de 30s.
func NewCustomerHttpClient(baseURL, serviceToken string) *CustomerHttpClient {
	return &CustomerHttpClient{
		BaseURL:      baseURL,
		ServiceToken: serviceToken,
		HTTPClient:   &http.Client{Timeout: 30 * time.Second},
	}
}

// ListAll lista todos os clientes cadastrados na API legada.
func (c *CustomerHttpClient) ListAll(ctx context.Context) ([]model.Customer, error) {
	var customers []model.Customer
	if err := c.doJSON(ctx, http.MethodGet, "/customers", nil, &customers); err != nil {
		return nil, fmt.Errorf("falha ao listar clientes da API legada: %w", err)
	}
	return customers, nil
}

// FindByID busca um cliente específico pelo ID. Retorna nil se não
// encontrado ou se ocorrer qualquer erro (mesmo comportamento "silencioso"
// de CustomerHttpClient.findById, que apenas loga e retorna null).
func (c *CustomerHttpClient) FindByID(ctx context.Context, id string) *model.Customer {
	var customer model.Customer
	if err := c.doJSON(ctx, http.MethodGet, "/customers/"+id, nil, &customer); err != nil {
		return nil
	}
	return &customer
}

// Create cria um novo cliente na API legada.
func (c *CustomerHttpClient) Create(ctx context.Context, name, phone string) model.MutationResult {
	payload := map[string]string{"name": name, "phone": phone}

	var created model.Customer
	if err := c.doJSON(ctx, http.MethodPost, "/customers", payload, &created); err != nil {
		return model.MutationResult{Success: false, Message: "Falha ao criar cliente: " + err.Error()}
	}
	return model.MutationResult{Success: true, Message: "Cliente criado com sucesso", Customer: &created}
}

// Update atualiza os dados de um cliente existente. name/phone vazios não
// são enviados no corpo (equivalente ao payload parcial da versão Java).
func (c *CustomerHttpClient) Update(ctx context.Context, id, name, phone string) model.MutationResult {
	payload := map[string]string{}
	if name != "" {
		payload["name"] = name
	}
	if phone != "" {
		payload["phone"] = phone
	}
	if len(payload) == 0 {
		return model.MutationResult{Success: false, Message: "Nenhum campo fornecido para atualização"}
	}

	var updated model.Customer
	if err := c.doJSON(ctx, http.MethodPut, "/customers/"+id, payload, &updated); err != nil {
		return model.MutationResult{Success: false, Message: "Falha ao atualizar cliente: " + err.Error()}
	}
	return model.MutationResult{Success: true, Message: "Cliente atualizado com sucesso", Customer: &updated}
}

// Delete remove um cliente da API legada pelo ID.
func (c *CustomerHttpClient) Delete(ctx context.Context, id string) model.MutationResult {
	if err := c.doJSON(ctx, http.MethodDelete, "/customers/"+id, nil, nil); err != nil {
		return model.MutationResult{Success: false, Message: "Falha ao remover cliente: " + err.Error()}
	}
	return model.MutationResult{Success: true, Message: "Cliente removido com sucesso"}
}

// doJSON executa uma requisição HTTP com corpo JSON opcional e desserializa
// a resposta em out (se não nil).
func (c *CustomerHttpClient) doJSON(ctx context.Context, method, path string, body any, out any) error {
	var reqBody io.Reader
	if body != nil {
		payload, err := json.Marshal(body)
		if err != nil {
			return err
		}
		reqBody = bytes.NewReader(payload)
	}

	req, err := http.NewRequestWithContext(ctx, method, c.BaseURL+path, reqBody)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+c.ServiceToken)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("status %d: %s", resp.StatusCode, string(respBody))
	}

	if out == nil {
		return nil
	}
	return json.NewDecoder(resp.Body).Decode(out)
}
